v2/pkg/engine/resolve/loader.go

```diff
@@ -412,31 +412,45 @@ func (l *Loader) mergeResult(res *result, items []int) error {
 		return nil
 	}
 	if res.out.Len() == 0 {
-		return l.renderErrorsFailedToFetch(res, failedToFetchEmptyResponse)
+		return l.renderErrorsFailedToFetch(res, emptyGraphQLResponse)
 	}
-	node, err := l.data.AppendAnyJSONBytes(res.out.Bytes())
+
+	// We expect a JSON object according to GraphQL over HTTP spec
+	node, err := l.data.AppendObject(res.out.Bytes())
 	if err != nil {
-		return l.renderErrorsFailedToFetch(res, failedToFetchInvalidJSON)
+		return l.renderErrorsFailedToFetch(res, invalidGraphQLResponse)
 	}
 
-	// error handling
+	hasErrors := false
+
+	// We check if the subgraph response has errors
 	if res.postProcessing.SelectResponseErrorsPath != nil {
-		// look for errors in the response and merge them into the errors array
 		ref := l.data.Get(node, res.postProcessing.SelectResponseErrorsPath)
 		if ref != -1 {
+			hasErrors = l.data.NodeIsDefined(ref) && len(l.data.Nodes[ref].ArrayValues) > 0
+			// Look for errors in the response and merge them into the errors array
 			err = l.mergeErrors(res, ref)
 			if err != nil {
 				return errors.WithStack(err)
 			}
 		}
 	}
+
+	// We also check if any data is there to processed
 	if res.postProcessing.SelectResponseDataPath != nil {
 		node = l.data.Get(node, res.postProcessing.SelectResponseDataPath)
+		// Check if the data is an array and not null
```

Comment:
This comment is wrong I think.
We should check if data is defined and if it's an object.

Reply:
I updated it. The response is validated before https://github.com/wundergraph/graphql-go-tools/pull/776/files#diff-4a656e1687222772aeae74410688ea6636143a823951dff49c01736b72294bd2R419